goethe/corpora.py

```diff
@@ -1,168 +1,52 @@
-        strides = a.strides + (a.strides[-1],)
-        return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
-
-    def number_ngrams(self, n):
-        for numbers in self.number_sentences():
-            for ngram in self.rolling_window(np.array(numbers), n):
-                yield ngram
-
-    def number_ngrams_(self, n):
-        for numbers in self.number_sentences():
-            for i in range(0, len(numbers) - n + 1):
-                yield numbers[i:i+n]
-
-    def sentences(self, words=True):
-        """Iterate over lines in files returning sentences.
+        path = self.file_path(use_tokens=True)
+        with open(path) as f:
+            yield from (l.strip().split() for l in f)
+
+    def file_path(self, use_tokens):
+        """Return path to either sentence or token file.
+        Example:
+            'abc/def/corpus'
+                --> 'abc/def/corpus/corpus.txt'
+                --> 'abc/def/corpus/corpus.tokens.txt'
         """
-        files = self.sentence_files()
-        trans = str.maketrans('', '', self.tokenizer_filter)
-        for f in files:
-            for line in open(f):
-                # Lines are of form: 'LineNumber\tActualSentence\n'
-                sent = line.split('\t')[1].strip().lower().translate(trans)
-                yield word_tokenize(sent) if words else sent
-
-    def sentences_n(self, n):
-        for i, s in enumerate(self.sentences()):
-            if self.max_sentences and i >= n:
-                break
-            yield s
+        corpus_name = os.path.basename(self.path)
+        file = ('%s.tokens.txt' if use_tokens else '%s.txt') % corpus_name
+        return os.path.join(self.path, file)
 
-    def sentence_files(self):
-        """Find all sentence files in 'dirname'.
-        Assumes file tree like following:
-            dirname/
-                deu_news_2015_3M/
-                    deu_news_2015_3M-sentences.txt
-                    ...
-                deu_wikipedia_2014_3M/
-                    deu_wikipedia_2014_3M-sentences.txt
-                    ...
-                ...
+    def random(self, k, use_tokens=None):
+        """Randomly select a list of k elements from either tokens or sents.
+        (Will load k elements into memory!)
         """
-        files = []
-        for corpus in os.listdir(self.dir):
-            # Select only corpora with 'self.lang' language
-            if not corpus.startswith(self.lang):
-                continue
-
-            if self.corpus_name and corpus != self.corpus_name:
-                continue
-
-            for f in os.listdir(os.path.join(self.dir, corpus)):
-                if not f.endswith('sentences.txt'):
-                    continue
-                if self.name_filter and not self.name_filter(f):
-                    continue
-                sentences_file = os.path.join(self.dir, corpus, f)
-                files.append(sentences_file)
-        return files
+        iterator = (self
```

Comment:
I think there is a good reason to randomly sample data and not just use `limit`, especially as you can take new random samples in each epoch. Regarding memory: 35M sentences is just 4Gigs. So holding 10M sentences in Memory is a little more than 1Gig. I don't see a problem with that.

The most elegant way that I see is to have the corpus iterator behave like an iterator. Meaning it's implementing `__iter__` and not writing `corpus.tokens()` every time. And behave like the corpora that come with gensim (which also implement `__iter__`). You said you don't want to define two objects, e.g. `sentences = Corpus(use_tokens=False)` and `tokens = Corpus(use_tokens=False)`. That is the reason we have this structure. I personally think it's confusing that one iterator object can iterate in two ways over it's lifetime. And I think objects that iterate should reflect how they iterate in their name, i.e., define two objects.

Being able to pass `limit` and `random` to both `sentences` and `tokens` means you have the same sampling logic in both methods. If the user passes `limit` but not `random`, then you're keeping track of an iterator `i` and check if `i == limit`. You care about `i` only in that case. If he passes `random` or omits `limit` checking `i` in every iteration is unnecessary. If the user passes `limit` and `random`, `sentences()` calls `random()`. But then `random()` still needs all sentences and has to call `sentences()` back. Actually `random` doesn't know if it was called by `tokens()` or `sentences()`, that's why you still have to pass `use_tokens`. If you don't want to write `corpus.random(use_tokens=True)` every time, you still need the confusing `if else` in it.

After adding code to check `limit`, keep track of `i` sometimes, and maybe call `random` some other times, the elegant three lines of code in `sentences()` and `tokens()` are wrapped in confusing if else constructs and the lines of code double at least. In **both** methods.